Fixes cat backward formula to return correct gradient values for R -> C case

anjali411 opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/51627
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51681 Fixes cat backward formula to return correct gradient values for R -> C case**

Differential Revision: [D26238748](https://our.internmc.facebook.com/intern/diff/D26238748)

anjali411 committed:
Fixes cat backward formula to return correct gradient values for R -> C case

[ghstack-poisoned]

gchanan commented on test/test_autograd.py:
```diff
@@ -7787,6 +7787,16 @@ def backward(ctx, *grad):
+        def fn(x1, x2):
+            return torch.cat((x1, x2), dim=-1)
+
+        torch.autograd.gradcheck(fn, [inp_r, inp_c])
```
nit: do the other order too.

anjali411 replied:
the order doesn't matter though ...

anjali411 replied:
anyway added the test for the other order, just in case

anjali411 committed:
Update on "Fixes cat backward formula to return correct gradient values for R -> C case"

Fixes https://github.com/pytorch/pytorch/issues/51627

Differential Revision: [D26238748](https://our.internmc.facebook.com/intern/diff/D26238748)

[ghstack-poisoned]

anjali411 committed:
Update on "Fixes cat backward formula to return correct gradient values for R -> C case"

Fixes https://github.com/pytorch/pytorch/issues/51627

Differential Revision: [D26238748](https://our.internmc.facebook.com/intern/diff/D26238748)

[ghstack-poisoned]

gchanan commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -553,23 +553,25 @@ Tensor unsqueeze_to(const Tensor & self, int64_t dim, IntArrayRef sizes) {
+
   for (size_t i = 0; i < sizes.size(); ++i) {
     auto& shape = sizes[i];
+    auto grad_ = handle_r_to_c(dtypes[i], grad);
```
I guess you don't really need to do this each time through (i.e. you only need at most 1 real view of the grad).  But probably doesn't matter performance wise.

albanD reviewed (approved):
LGTM

anjali411 committed:
Update on "Fixes cat backward formula to return correct gradient values for R -> C case"

Fixes https://github.com/pytorch/pytorch/issues/51627

Differential Revision: [D26238748](https://our.internmc.facebook.com/intern/diff/D26238748)

[ghstack-poisoned]
